### DISCUSSION TITLE
socket.io.on causes crash because io does not exist

### DISCUSSION THREAD

**Discussion by gperrow** [socketio/socket.io#3801] [Q&A] [answered]
I'm porting my application from socket.io 2.3.0 to 3.1.1. The [porting docs](https://socket.io/docs/v3/migrating-from-2-x-to-3-0/) as well as the [socket instance docs](https://socket.io/docs/v3/client-socket-instance/) both say that some of the events I used to listen on have been moved, so `socket.on( 'reconnect_attempt', func );` should now be `socket.io.on( 'reconnect_attempt', func );`. But in my 'connection' handler function, attempting to set these callback functions fails because the socket instance has no property called 'io'. Thus `socket.io.on` crashes with "Cannot read property 'on' of undefined".

Am I misunderstanding the change?

**Comment by darrachequesne** [comment, marked as answer]
> But in my 'connection' handler function

I think you are referring to the server Socket instance, not the client Socket instance. On the server-side, `socket.io` indeed does not exist.

Documentation:

- https://socket.io/docs/v3/server-socket-instance/#Events
- https://socket.io/docs/v3/client-socket-instance/#Events

**Reply by gperrow** [reply]
Of course. I should have seen that. Thanks
